### DISCUSSION TITLE
How to use zfs receive for encrypted datasets?

### DISCUSSION THREAD

**Discussion by hhb777** [openzfs/zfs#15158] [Q&A] [answered]
My steps:
1.Create a Encrypting File System as the source, and create a snapshot.
2.Create another Encrypting File System on the receiving side.
When executing the zfs send and zfs receive commands, zfs prompts me cannot receive new filesystem stream: destination dataset exists must specify -F to overwrite it.When I add the - F parameter, it prompts me cannot receive new filesystem stream: zfs receive -F cannot be used to destroy an encrypted filesystem or overwrite an unencryoted one with an encrypted one.
How to achieve zfs receive when the target dataset must exist and be encrypted? Or a bug in zfs?
Looking forward to a reply, thank you very much.

**Comment by rincebrain** [comment]
Don't try to receive by destroying something on the receiver side?

You appear to be trying to do something like
```
zfs create -o encryption=on pool1/source
zfs create -o encryption=on pool2/alreadyexists
zfs snapshot pool1/source@snap1
zfs send -w pool1/source@snap1 | zfs receive -F pool2/alreadyexists
```

When what you would want to do is
```
zfs create -o encryption=on pool1/source
zfs snapshot pool1/source@snap1
zfs send -w pool1/source@snap1 | zfs receive pool2/doesnotexist
```

**Reply by hhb777** [reply]
I would like to use the - F parameter on the receiver side.
I tried the method you provided, Unfortunately, zfs still prompted me cannot be used to destroy an encrypted filesystem or overwrite an unencryoted one with an encrypted one.
`zfs create -o encryption=on -o keyformat=passphrase pool1/source `
`zfs create -o encryption=on  -o keyformat=passphrase pool2/target`
`zfs snapshot pool1/source@snap1`
`zfs send -w pool1/source@snap1 | zfs receive -s -F pool2/target`
` cannot receive new filesystem stream: zfs receive -F cannot be used to destroy an encrypted filesystem or overwrite an unencrypted one with an encrypted one.`

The version of zfs I installed is **zfs-2.0.7**.
Do I need to upgrade the version in order to use - F on the encrypted dataset receiver?

**Reply by rincebrain** [reply]
You cannot destroy the destination in place when using native encryption, on any version.

You haven't explained why you want this, though.

**Reply by hhb777** [reply]
Sorry, I didn't explain my needs clearly.
I want on the receiver side to have an existing encrypted dataset, and to force the receiver side's encrypted dataset to be consistent with the source dataset during the first zfs receive

**Reply by rincebrain** [reply]
I don't understand what you mean here, I think.

That receive wouldn't do that, it just deletes whatever was there in the first place, no matter what it was. Having a dataset there already serves no purpose.

**Reply by nonanonymousanon** [reply, marked as answer]
I know this is exceptionally old but I suspect I understand what hhb777 is trying to accomplish since I was trying to do the same thing. 

You will want to destroy the existing encrypted dataset 'pool2/target' and then do a send and receive using override flags if youre trying to change the passphrase or key format or some other parameter as you move the pool. 

As an example:

`zfs create -o encryption=on -o keyformat=passphrase pool1/source `

`zfs snapshot pool1/source@snap1`
`zfs send -w pool1/source@snap1 | zfs receive -s  -o encryption=on -o keyformat=passphrase pool2/target`

**Reply by hhb777** [reply]
Yeah, I think you're right because I did the same thing. Thank you very much
